ui/src/app/shared/use-editable-resource.ts

```diff
@@ -0,0 +1,22 @@
+import {useState} from 'react';
+
+/**
+ * useEditableResource is a hook to manage the state of a resource that be edited and updated.
+ */
```

Comment:
Since this is JSDoc and not GoDoc, we can simplify this to be less repetitive / verbose. There's also a typo "that be edited"

We should also mention the ref comparison same as we did the JSON comparison in the `isEqual` function before. And can one-line it
```suggestion
/** Hook that manages the state of a resource that can be edited. Uses ref comparison */
```

Reply:
It also could make sense to rename this function/file `use-editable-object` to match `object-editor`.

Maybe even makes sense to have as part of the `ObjectEditor` component? 🤔

Reply:
The other hooks that have documentation follow the `<x> is a React hook` format:
https://github.com/argoproj/argo-workflows/blob/5f70eee137344f98786f61168cb32cbe7e493ec0/ui/src/app/shared/use-resizable-width.ts#L27
https://github.com/argoproj/argo-workflows/blob/5f70eee137344f98786f61168cb32cbe7e493ec0/ui/src/app/shared/use-resizable-width.ts#L27

I updated the comment in cf9a3f383d51dc74fc6e8edd0f86b9f413e14d52 to match, and renamed it to `use-editable-object`.

I don't see how this could be part of `<ObjectEditor>` because the `edited` variable is needed by the details components.

Reply:
> The other hooks that have documentation follow the `<x> is a React hook` format:

Ah that works for consistency then I suppose. Some of the original ones written were almost certainly bringing GoDoc conventions into JSDoc (there's a handful of non-JS idioms in the UI here and there).

> I don't see how this could be part of `<ObjectEditor>` because the `edited` variable is needed by the details components.

I meant as another exported function inside the same file, not directly integrated into the component.
As in, there are some conventions in the ecosystem of `import {useComponent, Component}` that we could follow. Although the usage is a tad different here with regard to the tracked state and props

Reply:
@MasonM any thoughts on the above? Otherwise this LGTM and I can merge

Reply:
I don't think it makes sense to couple this to `<ObjectEditor>`, because that will make it more difficult to move forward with the approach I posted above in https://github.com/argoproj/argo-workflows/pull/13593#discussion_r1781972661. That POC fully decouples `<ObjectEditorPlain>` from all serialization/deseralization and moves that logic into `useEditableComponent()`, which significantly improves performance and is a bit cleaner (IMO). Of course, that's a fairly significant change, and probably should come after this PR, but I think we'll want to do this eventually.

But if you feel strongly, I can go ahead and make that change.

Reply:
I don't feel strongly.
I didn't read that change too closely yet so from a quick look I thought putting the imports/files together might actually help move in that direction, but I might've been mistaken. Since the performance change may make it even more likely that any consumer of the component would want to also use the hook (and it is currently used like that) and so combining the imports makes it more obvious how to use.

> `useEditableComponent()`

I guess your intent with this naming is that the hook could in the future be extended to other editable components as well? As in that it's not specific to `ObjectEditor` per se

Reply:
> I guess your intent with this naming is that the hook could in the future be extended to other editable components as well?

Honestly, that was a typo. I meant `useEditableObject()`. It could theoretically be extended to other editable components, but I can't think of anything that would benefit.

Reply:
Oh ok, then I think it would make sense to combine the imports / exports so that it's more obvious the hook is supposed to be used together with the component. 

Technically doesn't have to be in the same files, could have an import/export stub instead